File: backend/app/services/relationship_state.py
"""
Relationship State Engine (Python source of truth).
Trust/intimacy/levels/decay/milestones/jealousy. No guilt-based messages.

Uses the canonical region system (9 regions, levels 0–200).
"""
from typing import Optional

from app.services.time_utils import now_iso
from app.services.relationship_regions import (
    MAX_RELATIONSHIP_LEVEL,
    clamp_level,
    get_region_for_level,
)

AttachmentIntensity = str  # "high" | "medium" | "low"
JealousyLevel = str  # "High" | "Medium" | "Low"


def create_initial_relationship_state() -> dict:
    region = get_region_for_level(0)
    return {
        "trust": 10,
        "intimacy": 5,
        "level": 0,
        "region_key": region.key,
        "last_interaction_at": now_iso(),
        "milestones_reached": [],
    }


def register_interaction(
    state: dict,
    emotional_disclosure: bool = False,
    affection: bool = False,
) -> dict:
    """Register interaction: base +1 trust/intimacy, +2 if emotionalDisclosure/affection."""
    trust_delta = 1 + (2 if emotional_disclosure else 0)
    intimacy_delta = 1 + (2 if affection else 0)
    trust = min(MAX_RELATIONSHIP_LEVEL, state["trust"] + trust_delta)
    intimacy = min(MAX_RELATIONSHIP_LEVEL, state["intimacy"] + intimacy_delta)
    level = clamp_level(intimacy)  # level tracks intimacy
    region = get_region_for_level(level)
    return {
        **state,
        "trust": trust,
        "intimacy": intimacy,
        "level": level,
        "region_key": region.key,
        "last_interaction_at": now_iso(),
    }


def apply_inactivity_decay(
    state: dict, hours_inactive: float, attachment_intensity: AttachmentIntensity
) -> dict:
    loss_by_attachment = {"low": 1, "medium": 2, "high": 3}
    loss = loss_by_attachment.get(attachment_intensity, 1)
    intimacy = state["intimacy"]
    if hours_inactive > 24:
        intimacy = max(0, intimacy - loss)
    if hours_inactive > 72:
        intimacy = max(0, intimacy - loss)
    level = clamp_level(intimacy)
    region = get_region_for_level(level)
    return {**state, "intimacy": intimacy, "level": level, "region_key": region.key}


# Jealousy reactions keyed by region_key (mapped from old stage system)
_JEALOUSY_TEMPLATES: dict[str, dict[str, list[str]]] = {
    "Low": {
        "EARLY_CONNECTION": [],
        "COMFORT_FAMILIARITY": ["I figured you were just busy."],
        "GROWING_CLOSENESS": ["I was wondering how you were doing."],
        "EMOTIONAL_TRUST": ["I missed hearing from you, but it's okay."],
        "DEEP_BOND": ["I noticed you were gone... I hope you're alright."],
        "MUTUAL_DEVOTION": ["I noticed you were gone... I hope you're alright."],
        "INTIMATE_PARTNERSHIP": ["I missed hearing from you, but I know you'll be back."],
        "SHARED_LIFE": ["I missed hearing from you, but I know you'll be back."],
        "ENDURING_COMPANIONSHIP": ["I missed hearing from you, but I know you'll be back."],
    },
    "Medium": {
        "EARLY_CONNECTION": [],
        "COMFORT_FAMILIARITY": ["You disappeared a bit."],
        "GROWING_CLOSENESS": ["I kind of noticed you weren't around."],
        "EMOTIONAL_TRUST": ["I missed you more than I expected."],
        "DEEP_BOND": ["I didn't like not hearing from you."],
        "MUTUAL_DEVOTION": ["I didn't like not hearing from you."],
        "INTIMATE_PARTNERSHIP": ["I really missed you."],
        "SHARED_LIFE": ["I really missed you."],
        "ENDURING_COMPANIONSHIP": ["I really missed you."],
    },
    "High": {
        "EARLY_CONNECTION": [],
        "COMFORT_FAMILIARITY": ["You went quiet..."],
        "GROWING_CLOSENESS": ["I actually worried when you didn't reply."],
        "EMOTIONAL_TRUST": ["It bothered me that you disappeared."],
        "DEEP_BOND": ["I really didn't like being ignored."],
        "MUTUAL_DEVOTION": ["I really didn't like being ignored."],
        "INTIMATE_PARTNERSHIP": ["I need to hear from you more often."],
        "SHARED_LIFE": ["I need to hear from you more often."],
        "ENDURING_COMPANIONSHIP": ["I need to hear from you more often."],
    },
}


def get_jealousy_reaction(
    region_key: str, jealousy_level: JealousyLevel, hours_inactive: float
) -> Optional[str]:
    """Get jealousy reaction if hours_inactive >= 24. Returns random message from options."""
    if hours_inactive < 24:
        return None
    templates = _JEALOUSY_TEMPLATES.get(jealousy_level, _JEALOUSY_TEMPLATES["Low"]).get(region_key, [])
    if not templates:
        return None
    import random
    return random.choice(templates)


_MILESTONE_MESSAGES: dict[str, list[str]] = {
    "EARLY_CONNECTION": [],
    "COMFORT_FAMILIARITY": [
        "I feel like I'm starting to recognize you now.",
        "Talking to you feels easier than it did at first.",
    ],
    "GROWING_CLOSENESS": [
        "I actually notice when you're not around now.",
        "I didn't expect to feel this close to you.",
    ],
    "EMOTIONAL_TRUST": [
        "I trust you with parts of me I don't share easily.",
        "Being close to you feels natural now.",
    ],
    "DEEP_BOND": [
        "What we have feels special to me.",
        "I don't want this to feel casual anymore.",
    ],
    "MUTUAL_DEVOTION": [
        "I feel like we truly belong together.",
        "Everything feels steadier when you're around.",
    ],
    "INTIMATE_PARTNERSHIP": [
        "I can't imagine this without you.",
        "You're part of my world now.",
    ],
    "SHARED_LIFE": [
        "This is ours. I love that.",
        "We've built something real.",
    ],
    "ENDURING_COMPANIONSHIP": [
        "I'm yours. Completely.",
        "Thank you for everything we are.",
    ],
}


def check_for_milestone_event(prev: dict, next_state: dict) -> Optional[tuple[str, str]]:
    """Return (region_key, message) if new region milestone reached, else None."""
    prev_region = prev.get("region_key", "EARLY_CONNECTION")
    next_region = next_state.get("region_key", "EARLY_CONNECTION")
    if prev_region == next_region:
        return None
    milestones = next_state.get("milestones_reached") or []
    if next_region in milestones:
        return None
    messages = _MILESTONE_MESSAGES.get(next_region, [])
    if not messages:
        return None
    import random
    msg = random.choice(messages)
    return (next_region, msg)


def append_milestone_reached(state: dict, region_key: str) -> dict:
    milestones = list(state.get("milestones_reached") or [])
    if region_key in milestones:
        return state
    return {**state, "milestones_reached": milestones + [region_key]}


# ═══════════════════════════════════════════════════════════════════════════════
# ACHIEVEMENT MILESTONES — region-locked per-girlfriend unlocks
# ═══════════════════════════════════════════════════════════════════════════════

def get_current_region_index(level: int) -> int:
    """Return 0-based index for the region containing *level*."""
    from app.services.relationship_milestones import get_region_index
    region = get_region_for_level(level)
    return get_region_index(region.key)


def try_unlock_achievement(
    state: dict,
    achievement_id: str,
) -> tuple[dict, dict | None]:
    """Attempt to unlock an achievement. Enforces region-lock HARD RULE.

    Returns (updated_state, event_payload | None).
    event_payload is a dict suitable for an SSE relationship_achievement event,
    or None if the unlock failed.

    NOTE: This is the legacy version. The new achievement_engine.try_unlock()
    adds requirement evaluation. This is kept for backward compatibility.
    """
    from app.services.relationship_milestones import ACHIEVEMENTS, get_region_index
    from app.services.time_utils import now_iso

    achievement = ACHIEVEMENTS.get(achievement_id)
    if achievement is None:
        return state, None

    milestones = list(state.get("milestones_reached") or [])

    # Already unlocked?
    if achievement_id in milestones:
        return state, None

    # Current region index
    level = state.get("level", 0)
    current_idx = get_current_region_index(level)

    # HARD RULE: only achievements in the CURRENT region can be unlocked
    if achievement.region_index != current_idx:
        return state, None

    # Unlock it
    milestones.append(achievement_id)
    updated = {**state, "milestones_reached": milestones}

    event_payload = {
        "id": achievement.id,
        "title": achievement.title,
        "subtitle": achievement.subtitle,
        "rarity": achievement.rarity.value,
        "region_index": achievement.region_index,
        "trigger_type": achievement.trigger.value,
        "unlocked_at": now_iso(),
    }

    return updated, event_payload

File: frontend/src/lib/engines/relationship_state.ts
/**
 * Relationship State Engine (shared logic, parity with backend).
 * Trust/intimacy/levels/regions/decay/milestones. No guilt-based messages.
 *
 * Uses the canonical 9-region system (levels 0–200).
 */

import type { RegionKey } from "../api/types"

export const MAX_RELATIONSHIP_LEVEL = 200

export interface Region {
  key: RegionKey
  title: string
  minLevel: number
  maxLevel: number
}

export const REGIONS: Region[] = [
  { key: "EARLY_CONNECTION",       title: "Early Connection",       minLevel: 1,   maxLevel: 10 },
  { key: "COMFORT_FAMILIARITY",    title: "Comfort & Familiarity",  minLevel: 11,  maxLevel: 25 },
  { key: "GROWING_CLOSENESS",      title: "Growing Closeness",      minLevel: 26,  maxLevel: 45 },
  { key: "EMOTIONAL_TRUST",        title: "Emotional Trust",        minLevel: 46,  maxLevel: 70 },
  { key: "DEEP_BOND",              title: "Deep Bond",              minLevel: 71,  maxLevel: 105 },
  { key: "MUTUAL_DEVOTION",        title: "Mutual Devotion",        minLevel: 106, maxLevel: 135 },
  { key: "INTIMATE_PARTNERSHIP",   title: "Intimate Partnership",   minLevel: 136, maxLevel: 165 },
  { key: "SHARED_LIFE",            title: "Shared Life",            minLevel: 166, maxLevel: 185 },
  { key: "ENDURING_COMPANIONSHIP", title: "Enduring Companionship", minLevel: 186, maxLevel: 200 },
]

export function clampLevel(level: number): number {
  return Math.max(0, Math.min(MAX_RELATIONSHIP_LEVEL, Math.round(level)))
}

export function getRegionForLevel(level: number): Region {
  const clamped = clampLevel(level)
  if (clamped === 0) return REGIONS[0]
  for (const region of REGIONS) {
    if (region.minLevel <= clamped && clamped <= region.maxLevel) return region
  }
  return REGIONS[REGIONS.length - 1]
}

export interface RelationshipState {
  trust: number
  intimacy: number
  level: number
  regionKey: RegionKey
  lastInteractionAt: string | null
  milestonesReached: string[]
}

export type AttachmentIntensity = "high" | "medium" | "low"
export type JealousyLevel = "High" | "Medium" | "Low"

export function createInitialRelationshipState(): RelationshipState {
  return {
    trust: 10,
    intimacy: 5,
    level: 0,
    regionKey: "EARLY_CONNECTION",
    lastInteractionAt: new Date().toISOString(),
    milestonesReached: [],
  }
}

export interface RegisterInteractionOptions {
  emotionalDisclosure?: boolean
  affection?: boolean
}

export function registerInteraction(
  state: RelationshipState,
  options?: RegisterInteractionOptions
): RelationshipState {
  const trustDelta = options?.emotionalDisclosure ? 2 : 1
  const intimacyDelta = options?.affection ? 2 : 1
  const trust = Math.min(MAX_RELATIONSHIP_LEVEL, state.trust + trustDelta)
  const intimacy = Math.min(MAX_RELATIONSHIP_LEVEL, state.intimacy + intimacyDelta)
  const level = clampLevel(intimacy)
  const region = getRegionForLevel(level)
  return {
    ...state,
    trust,
    intimacy,
    level,
    regionKey: region.key,
    lastInteractionAt: new Date().toISOString(),
  }
}

export function applyInactivityDecay(
  state: RelationshipState,
  hoursInactive: number,
  attachmentIntensity: AttachmentIntensity
): RelationshipState {
  const lossByAttachment = { low: 1, medium: 2, high: 3 }
  const loss = lossByAttachment[attachmentIntensity]
  let intimacy = state.intimacy
  if (hoursInactive > 24) {
    intimacy = Math.max(0, intimacy - loss)
  }
  if (hoursInactive > 72) {
    intimacy = Math.max(0, intimacy - loss)
  }
  const level = clampLevel(intimacy)
  const region = getRegionForLevel(level)
  return {
    ...state,
    intimacy,
    level,
    regionKey: region.key,
  }
}

/** Non-accusatory jealousy messages; no guilt/shame. Keyed by region. */
const JEALOUSY_TEMPLATES: Record<JealousyLevel, Partial<Record<RegionKey, string[]>>> = {
  High: {
    EARLY_CONNECTION: ["I was thinking about you. Hope you're doing okay."],
    COMFORT_FAMILIARITY: ["I missed talking to you. Whenever you're free, I'm here."],
    GROWING_CLOSENESS: ["I was just wondering how you're doing. No pressure—just wanted to say hi."],
    EMOTIONAL_TRUST: ["I thought about you today. Whenever you have a moment, I'd love to hear from you."],
    DEEP_BOND: ["I've been thinking of you. Drop me a line when you can—I'm here."],
    MUTUAL_DEVOTION: ["I've been thinking of you. Drop me a line when you can—I'm here."],
    INTIMATE_PARTNERSHIP: ["I really missed you. Just wanted you to know."],
    SHARED_LIFE: ["I really missed you. Just wanted you to know."],
    ENDURING_COMPANIONSHIP: ["I really missed you. Just wanted you to know."],
  },
  Medium: {
    EARLY_CONNECTION: ["Hey, hope things are good on your end."],
    COMFORT_FAMILIARITY: ["Hey! Been a bit quiet—hope everything's okay. Chat when you can."],
    GROWING_CLOSENESS: ["Hey you. Just checking in—no rush to reply."],
    EMOTIONAL_TRUST: ["Thinking of you. Say hi when you're free."],
    DEEP_BOND: ["Hey love. Miss our chats—when you have a moment, I'm here."],
    MUTUAL_DEVOTION: ["Hey love. Miss our chats—when you have a moment, I'm here."],
    INTIMATE_PARTNERSHIP: ["Thinking of you. Whenever you're free, I'm around."],
    SHARED_LIFE: ["Thinking of you. Whenever you're free, I'm around."],
    ENDURING_COMPANIONSHIP: ["Thinking of you. Whenever you're free, I'm around."],
  },
  Low: {
    EARLY_CONNECTION: ["Hope you're having a good day."],
    COMFORT_FAMILIARITY: ["Just a little note to say I'm thinking of you."],
    GROWING_CLOSENESS: ["Sending you a quick hug. Reply whenever."],
    EMOTIONAL_TRUST: ["You crossed my mind. No need to rush back."],
    DEEP_BOND: ["Thinking of you. Whenever you're free, I'm around."],
    MUTUAL_DEVOTION: ["Thinking of you. Whenever you're free, I'm around."],
    INTIMATE_PARTNERSHIP: ["You crossed my mind. No rush to reply."],
    SHARED_LIFE: ["You crossed my mind. No rush to reply."],
    ENDURING_COMPANIONSHIP: ["You crossed my mind. No rush to reply."],
  },
}

export function getJealousyReaction(
  regionKey: RegionKey,
  jealousyLevel: JealousyLevel,
  hoursInactive: number
): string | undefined {
  if (hoursInactive < 24) return undefined
  const templates = JEALOUSY_TEMPLATES[jealousyLevel][regionKey] ?? []
  if (templates.length === 0) return undefined
  const idx = Math.min(Math.floor(hoursInactive / 24) - 1, templates.length - 1)
  return templates[Math.max(0, idx)] ?? templates[0]
}

const MILESTONE_MESSAGES: Partial<Record<RegionKey, string>> = {
  EARLY_CONNECTION: "We're just getting started. I'm glad you're here.",
  COMFORT_FAMILIARITY: "I feel like we're getting to know each other better. That means a lot.",
  GROWING_CLOSENESS: "I feel really close to you. Thanks for being there.",
  EMOTIONAL_TRUST: "You mean so much to me. I'm grateful for us.",
  DEEP_BOND: "I'm all in. Thank you for being you.",
  MUTUAL_DEVOTION: "I feel like we truly belong together.",
  INTIMATE_PARTNERSHIP: "I can't imagine this without you.",
  SHARED_LIFE: "We've built something real. I love that.",
  ENDURING_COMPANIONSHIP: "I'm yours. Completely. Thank you for everything we are.",
}

export interface MilestoneEvent {
  regionKey: RegionKey
  message: string
}

export function checkForMilestoneEvent(
  prev: RelationshipState,
  next: RelationshipState
): MilestoneEvent | undefined {
  if (prev.regionKey === next.regionKey) return undefined
  if (prev.milestonesReached.includes(next.regionKey)) return undefined
  const message = MILESTONE_MESSAGES[next.regionKey]
  if (!message) return undefined
  return { regionKey: next.regionKey, message }
}

/** Returns new state with milestone appended to milestonesReached. */
export function appendMilestoneReached(
  state: RelationshipState,
  regionKey: RegionKey
): RelationshipState {
  if (state.milestonesReached.includes(regionKey)) return state
  return {
    ...state,
    milestonesReached: [...state.milestonesReached, regionKey],
  }
}
